File: app/src/main/java/nu/huw/clarity/db/model/NoteHelper.java
package nu.huw.clarity.db.model;

import android.util.Log;
import java.io.IOException;
import java.io.StringReader;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import org.w3c.dom.Document;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;

public class NoteHelper {

  private static final String TAG = NoteHelper.class.getSimpleName();

  public static String noteXMLtoString(String noteXML) {

    String string = "";

    try {

      DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
      DocumentBuilder db = dbf.newDocumentBuilder();
      Document document = db.parse(new InputSource(new StringReader(noteXML)));

      document.getDocumentElement().normalize();
      NodeList nodes = document.getFirstChild().getChildNodes();

      for (int i = 0; i < nodes.getLength(); i++) {

        Node node = nodes.item(i);

        if (node.getNodeType() == Node.ELEMENT_NODE && node.getNodeName().equals("text")) {

          // Nodes at this level should always be <p> paragraph nodes

          string = parseTextChildren(node, string);
        }
      }

    } catch (NullPointerException e) {
      Log.e(TAG, "Error with the XML format", e);
    } catch (ParserConfigurationException | SAXException | IOException e) {
      Log.e(TAG, "Error parsing XML", e);
    }

    return string;
  }

  private static String parseTextChildren(Node node, String string) {
    NodeList nodes = node.getChildNodes();
    for (int i = 0; i < nodes.getLength(); i++) {
      Node childNode = nodes.item(i);

      if (childNode.getNodeType() == Node.ELEMENT_NODE && childNode.getNodeName().equals("p")) {

        // Should also always be <p> nodes

        if (!string.isEmpty()) {
          string += "\n"; // add a newline
        }

        string = parseParagraphChildren(childNode, string);
      }
    }
    return string;
  }

  private static String parseParagraphChildren(Node node, String string) {
    NodeList nodes = node.getChildNodes();
    for (int i = 0; i < nodes.getLength(); i++) {
      Node childNode = nodes.item(i);

      if (childNode.getNodeType() == Node.ELEMENT_NODE && childNode.getNodeName().equals("run")) {

        // Should also always be <run> nodes as children of a <p>

        string = parseRunChildren(childNode, string);

      }
    }
    return string;
  }

  private static String parseRunChildren(Node node, String string) {
    NodeList nodes = node.getChildNodes();
    for (int i = 0; i < nodes.getLength(); i++) {
      Node childNode = nodes.item(i);
      if (childNode.getNodeType() == Node.ELEMENT_NODE && childNode.getNodeName().equals("lit")) {

        // Ignore <style> nodes, only <lit> nodes wanted

        string = parseLitChildren(childNode, string);

      }
    }
    return string;
  }

  /**
   * Most lit function in the entire app
   */
  private static String parseLitChildren(Node node, String string) {
    NodeList nodes = node.getChildNodes();
    for (int i = 0; i < nodes.getLength(); i++) {
      Node childNode = nodes.item(i);
      int nodeType = childNode.getNodeType();

      if (nodeType == Node.TEXT_NODE) {
        string += childNode.getNodeValue();
      }
    }
    return string;
  }

}

File: app/src/main/java/nu/huw/clarity/db/NoteHelper.kt
package nu.huw.clarity.db

import android.util.Log
import org.w3c.dom.Node
import org.xml.sax.InputSource
import org.xml.sax.SAXException
import java.io.IOException
import java.io.StringReader
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.parsers.ParserConfigurationException

object NoteHelper {

    private val TAG = NoteHelper::class.java.simpleName

    fun noteXMLtoString(noteXML: String): String {

        var string = ""

        try {

            val dbf = DocumentBuilderFactory.newInstance()
            val db = dbf.newDocumentBuilder()
            val document = db.parse(InputSource(StringReader(noteXML)))

            document.documentElement.normalize()
            val nodes = document.firstChild.childNodes

            (0..nodes.length - 1)
                    .asSequence()
                    .map { nodes.item(it) }
                    .filter { it.nodeType == Node.ELEMENT_NODE && it.nodeName == "text" }
                    .forEach {
                        // Nodes at this level should always be <p> paragraph nodes
                        string = parseTextChildren(it, string)
                    }

        } catch (e: NullPointerException) {
            Log.e(TAG, "Error with the XML format", e)
        } catch (e: ParserConfigurationException) {
            Log.e(TAG, "Error parsing XML", e)
        } catch (e: SAXException) {
            Log.e(TAG, "Error parsing XML", e)
        } catch (e: IOException) {
            Log.e(TAG, "Error parsing XML", e)
        }

        return string
    }

    private fun parseTextChildren(node: Node, _string: String): String {
        var string = _string
        val nodes = node.childNodes

        (0..nodes.length - 1)
                .asSequence()
                .map { nodes.item(it) }
                .filter { it.nodeType == Node.ELEMENT_NODE && it.nodeName == "p" }
                .forEach {
                    // Should also always be <p> nodes
                    if (!string.isEmpty()) {
                        string += "\n" // add a newline
                    }
                    string = parseParagraphChildren(it, string)
                }

        return string
    }

    private fun parseParagraphChildren(node: Node, _string: String): String {
        var string = _string
        val nodes = node.childNodes
        (0..nodes.length - 1)
                .asSequence()
                .map { nodes.item(it) }
                .filter { it.nodeType == Node.ELEMENT_NODE && it.nodeName == "run" }
                .forEach {
                    // Should also always be <run> nodes as children of a <p>

                    string = parseRunChildren(it, string)
                }
        return string
    }

    private fun parseRunChildren(node: Node, _string: String): String {
        var string = _string
        val nodes = node.childNodes
        (0..nodes.length - 1)
                .asSequence()
                .map { nodes.item(it) }
                .filter { it.nodeType == Node.ELEMENT_NODE && it.nodeName == "lit" }
                .forEach {
                    // Ignore <style> nodes, only <lit> nodes wanted

                    string = parseLitChildren(it, string)
                }
        return string
    }

    /**
     * Most lit function in the entire app
     */
    private fun parseLitChildren(node: Node, _string: String): String {
        var string = _string
        val nodes = node.childNodes
        (0..nodes.length - 1)
                .map { nodes.item(it) }
                .filter { it.nodeType.toInt() == Node.TEXT_NODE.toInt() }
                .forEach { string += it.nodeValue }
        return string
    }

}